### PR TITLE
[ADP-1963] split Tx module in preparation to augment TransactionInfo

### PR DESCRIPTION
**Pull request by paolino**
I have mostly splitted `Tx` module in parts
    - [x]  moved out `Tx` in its own module 
    - [x]  moved out `TransactionInfo` in its own module 
    - [x]  moved out `TxMeta` in its own module     
    - [x]  moved out `SealedTx` in its own module 
    - [x] moved token bundle assessments data types to `TokenBundle` module
    - [x]  moved out constraints in their own module 

### Comments

This is  not necessary for the ADP, but as I need to modify TransactionInfo and potentially TxMeta, I wanted to have some clarity on the dependencies

### Issue Number

ADP-1963

### CODE REVIEW

**Comment by paolino** [comment]
not ready sorry for clicking the button

**Comment by paolino** [comment]
CI failed:
```
  To rerun use: --match "/Cardano.Wallet.Api.Server.Tls/TLS Client Authentication/Deny client with wrong certificate if TLS is enabled/"

Randomized with seed 1197664393
```

**Review by HeinrichApfelmus — APPROVED** [review]
Excellent idea, thank you! 😊

File: lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs [hunk 1]
```diff
@@ -187,7 +188,8 @@ import qualified Cardano.Wallet.Primitive.Types.Coin as Coin
 import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
 import qualified Cardano.Wallet.Primitive.Types.TokenMap as TokenMap
 import qualified Cardano.Wallet.Primitive.Types.TokenQuantity as TokenQuantity
-import qualified Cardano.Wallet.Primitive.Types.UTxOIndex as UTxOIndex
+import qualified Cardano.Wallet.Primitive.Types.UTxOIndex.Internal as UTxOIndex
+    ( UTxOIndex (balance), selectRandomWithPriority, toList )
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs]: Hi @paolino 

According to our coding standards, [imports should be grouped into three groups, separated by a blank newline](https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md#stylish-haskell-is-used-to-format-grouped-imports--language-pragmas):

    Prelude import
    Explicit imports
    Qualified imports

So explicit imports like `UTxOIndex` should be moved to the section above.

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs [hunk 2]
```diff
@@ -4,19 +4,17 @@
...
+{-# LANGUAGE StrictData #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 
 -- |
 -- Copyright: © 2018-2020 IOHK
 -- License: Apache-2.0
 --
--- This module provides the main transaction data types used by the wallet.
+-- This module re-export all transaction types.
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs]:
```suggestion
-- This module re-exports all transaction types.
```

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Tx.hs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+txOutMapAssetIds :: (AssetId -> AssetId) -> TxOut -> TxOut
+txOutMapAssetIds f (TxOut address bundle) =
+    TxOut address (TokenBundle.mapAssetIds f bundle)
+
+txOutRemoveAssetId :: TxOut -> AssetId -> TxOut
+txOutRemoveAssetId (TxOut address bundle) asset =
+    TxOut address (TokenBundle.setQuantity bundle asset mempty)
+
+
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Tx.hs]: This pair of blank lines can be removed!

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Constraints.hs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- This module provides `TxConstraints` data type.
+--
+module Cardano.Wallet.Primitive.Types.Tx.Constraints
+    (
+    TxConstraints (..)
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx/Constraints.hs]:
```suggestion
    ( TxConstraints (..)
```

File: lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TransactionInfo.hs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- This module provides the `TransactionInfo` data types used by the wallet.
+--
+module Cardano.Wallet.Primitive.Types.Tx.TransactionInfo
+    where
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Primitive/Types/Tx/TransactionInfo.hs]: Maybe add a list of exports here?

File: lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs [hunk 6]
```diff
@@ -406,6 +407,8 @@ import qualified Cardano.Wallet.Primitive.Types.Coin as W
 import qualified Cardano.Wallet.Primitive.Types.Hash as W
 import qualified Cardano.Wallet.Primitive.Types.Redeemer as W
 import qualified Cardano.Wallet.Primitive.Types.RewardAccount as W
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundleSizeAssessment (..), TokenBundleSizeAssessor (..) )
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs]: Similar to above, according to our coding standards, [imports should be grouped into three groups, separated by a blank newline](https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md#stylish-haskell-is-used-to-format-grouped-imports--language-pragmas):

Prelude import
Explicit imports
Qualified imports

So explicit imports like `tokenBundleSizeAssessment` should be moved to the explicit imports section.

**Review by jonathanknowles — COMMENTED** [review]
Hi @paolino 

Just some quick comments!

One thing to bear in mind:

According to our coding standards, [imports should be grouped into three groups, separated by a blank newline](https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md#stylish-haskell-is-used-to-format-grouped-imports--language-pragmas):

- Prelude import
- Explicit imports
- Qualified imports

There are a few places in this PR where explicit imports and qualified imports are mixed up. Would it be possible to fix these? Many thanks!

**Comment by jonathanknowles** [comment]
Hi @paolino 

I have two other proposals:

- I would propose to move the `TokenBundleSize{Assessor,Assessment}` types outside of the `TokenBundle` module. The purpose of these types relates to token bundles **_in the context of transaction outputs_**, rather than token bundles per se. (Specifically, whether or not a token bundle is too large to fit in a transaction output.) So these size assessments are not integral to the definition of a token bundle.
    - Perhaps they could exist in their own module `Tx.TokenBundleSizeAssessment`. Placing this module under the `Tx` hierarchy clearly articulates that the context relates to transactions.
- I would propose to delete the `Primitive.Types.Tx` module (which now just re-exports things from elsewhere) and to re-arrange the imports in other modules to import directly from the specific modules they need. (Though perhaps in a second PR!)
    - I think re-exports are useful in some circumstances, but that there should be a clear benefit to their use that outweighs their disadvantages. Re-exports can make code harder to navigate, and can also lead to unnecessary compilation bottlenecks, particularly if other modules outside the hierarchy typically import from the "re-exporting" module instead of importing directly. See https://www.parsonsmatt.org/2019/11/27/keeping_compilation_fast.html)

**Comment by paolino** [comment]
> Hi @paolino
> 
> I have two other proposals:
> 
>     * I would propose to move the `TokenBundleSize{Assessor,Assessment}` types outside of the `TokenBundle` module. The purpose of these types relates to token bundles **_in the context of transaction outputs_**, rather than token bundles per se. (Specifically, whether or not a token bundle is too large to fit in a transaction output.) So these size assessments are not integral to the definition of a token bundle.
>       
>       * Perhaps they could exist in their own module `Tx.TokenBundleSizeAssessment`. Placing this module under the `Tx` hierarchy clearly articulates that the context relates to transactions.
> 
>     * I would propose to delete the `Primitive.Types.Tx` module (which now just re-exports things from elsewhere) and to re-arrange the imports in other modules to import directly from the specific modules they need. (Though perhaps in a second PR!)
>       
>       * I think re-exports are useful in some circumstances, but that there should be a clear benefit to their use that outweighs their disadvantages. Re-exports can make code harder to navigate, and can also lead to unnecessary compilation bottlenecks, particularly if other modules outside the hierarchy typically import from the "re-exporting" module instead of importing directly. See https://www.parsonsmatt.org/2019/11/27/keeping_compilation_fast.html)

Thanks for reviewing! 

I will move the TokenBundle assessments under Tx. 

I left the re-exporter to make the PR impact as small as possible. I'm no fan of re-exporting so I'm open to the big import refinement derived from the split. But 
  1. we should have a policy that helps deciding 
  1.  @HeinrichApfelmus in the `Read` hierarchy is going in the opposite direction
  1. fewer imports is also a value (even if re-exporting is not well supported by HLS import hints  :-( ) 
So I will leave the `Type.Tx` module for now.

**Pull request merged**